File: frontend/lib/geo.ts
import type { LngLat } from "./types";

const R = 6371; // km

export function haversineKm(a: LngLat, b: LngLat): number {
  const dLat = toRad(b[1] - a[1]);
  const dLng = toRad(a[0] - b[0]) * -1;
  const lat1 = toRad(a[1]);
  const lat2 = toRad(b[1]);
  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) ** 2;
  return 2 * R * Math.asin(Math.min(1, Math.sqrt(h)));
}

function toRad(d: number): number {
  return (d * Math.PI) / 180;
}

/** Total length of a polyline in km plus the cumulative distance per vertex. */
export function polylineCum(coords: LngLat[]): { total: number; cum: number[] } {
  const cum: number[] = [0];
  let total = 0;
  for (let i = 1; i < coords.length; i++) {
    total += haversineKm(coords[i - 1], coords[i]);
    cum.push(total);
  }
  return { total, cum };
}

/** Interpolate a position at distance `d` km along a polyline. Returns position + bearing. */
export function interpAlong(
  coords: LngLat[],
  cum: number[],
  d: number
): { pos: LngLat; bearing: number } {
  if (d <= 0) return { pos: coords[0], bearing: bearingBetween(coords[0], coords[1] ?? coords[0]) };
  const last = cum[cum.length - 1];
  if (d >= last)
    return {
      pos: coords[coords.length - 1],
      bearing: bearingBetween(coords[coords.length - 2] ?? coords[0], coords[coords.length - 1])
    };
  // binary-ish linear search
  let i = 1;
  while (i < cum.length && cum[i] < d) i++;
  const segStart = cum[i - 1];
  const segEnd = cum[i];
  const t = segEnd === segStart ? 0 : (d - segStart) / (segEnd - segStart);
  const a = coords[i - 1];
  const b = coords[i];
  const pos: LngLat = [a[0] + (b[0] - a[0]) * t, a[1] + (b[1] - a[1]) * t];
  return { pos, bearing: bearingBetween(a, b) };
}

export function bearingBetween(a: LngLat, b: LngLat): number {
  const y = Math.sin(toRad(b[0] - a[0])) * Math.cos(toRad(b[1]));
  const x =
    Math.cos(toRad(a[1])) * Math.sin(toRad(b[1])) -
    Math.sin(toRad(a[1])) * Math.cos(toRad(b[1])) * Math.cos(toRad(b[0] - a[0]));
  return (Math.atan2(y, x) * 180) / Math.PI;
}

export function fmtClock(sec: number): string {
  const s = ((sec % 86400) + 86400) % 86400;
  const h = Math.floor(s / 3600);
  const m = Math.floor((s % 3600) / 60);
  return `${String(h).padStart(2, "0")}:${String(m).padStart(2, "0")}`;
}

export function fmtClockS(sec: number): string {
  const s = ((sec % 86400) + 86400) % 86400;
  const h = Math.floor(s / 3600);
  const m = Math.floor((s % 3600) / 60);
  const ss = Math.floor(s % 60);
  return `${String(h).padStart(2, "0")}:${String(m).padStart(2, "0")}:${String(ss).padStart(2, "0")}`;
}

export function parseHHMM(t: string): number {
  const [h, m] = t.split(":").map(Number);
  return h * 3600 + m * 60;
}

File: backend/railmind/geo.py
"""Geodesy + arc-length helpers.

Arc-length parameterization is the foundation of smooth motion: positions are
addressed by distance travelled (km), not by vertex index, so visual speed is
correct regardless of how densely the polyline is sampled.
"""
from __future__ import annotations

import math

LngLat = tuple[float, float]
_R = 6371.0  # km


def haversine_km(a: LngLat, b: LngLat) -> float:
    lon1, lat1 = math.radians(a[0]), math.radians(a[1])
    lon2, lat2 = math.radians(b[0]), math.radians(b[1])
    dlat = lat2 - lat1
    dlon = lon2 - lon1
    h = math.sin(dlat / 2) ** 2 + math.cos(lat1) * math.cos(lat2) * math.sin(dlon / 2) ** 2
    return 2 * _R * math.asin(min(1.0, math.sqrt(h)))


def cumulative_arc_length(coords: list[LngLat]) -> list[float]:
    """Precomputed cumulative arc-length (km) at each vertex."""
    cum = [0.0]
    for i in range(1, len(coords)):
        cum.append(cum[-1] + haversine_km(coords[i - 1], coords[i]))
    return cum


def bearing_deg(a: LngLat, b: LngLat) -> float:
    lon1, lat1 = math.radians(a[0]), math.radians(a[1])
    lon2, lat2 = math.radians(b[0]), math.radians(b[1])
    dlon = lon2 - lon1
    y = math.sin(dlon) * math.cos(lat2)
    x = math.cos(lat1) * math.sin(lat2) - math.sin(lat1) * math.cos(lat2) * math.cos(dlon)
    return math.degrees(math.atan2(y, x))


def interpolate_along(coords: list[LngLat], cum: list[float], d: float) -> tuple[LngLat, float]:
    """Position + heading at distance ``d`` km along an arc-length polyline."""
    if d <= 0:
        nxt = coords[1] if len(coords) > 1 else coords[0]
        return coords[0], bearing_deg(coords[0], nxt)
    total = cum[-1]
    if d >= total:
        prev = coords[-2] if len(coords) > 1 else coords[-1]
        return coords[-1], bearing_deg(prev, coords[-1])
    # locate segment
    lo, hi = 0, len(cum) - 1
    while lo < hi - 1:
        mid = (lo + hi) // 2
        if cum[mid] <= d:
            lo = mid
        else:
            hi = mid
    seg_start, seg_end = cum[lo], cum[lo + 1]
    t = 0.0 if seg_end == seg_start else (d - seg_start) / (seg_end - seg_start)
    a, b = coords[lo], coords[lo + 1]
    pos = (a[0] + (b[0] - a[0]) * t, a[1] + (b[1] - a[1]) * t)
    return pos, bearing_deg(a, b)


def smootherstep(x: float) -> float:
    """Ken Perlin's smootherstep: zero 1st & 2nd derivative at the ends.

    Used as the distance-vs-time easing between two stops, giving smooth
    acceleration out of a station and deceleration into the next one.
    """
    x = max(0.0, min(1.0, x))
    return x * x * x * (x * (x * 6 - 15) + 10)


def smootherstep_deriv(x: float) -> float:
    """d/dx smootherstep — proportional to instantaneous speed along a segment."""
    x = max(0.0, min(1.0, x))
    return 30 * x * x * (x - 1) * (x - 1)


def parse_hhmm(t: str) -> int:
    h, m = t.split(":")
    return int(h) * 3600 + int(m) * 60
